rs/tests/driver/src/driver/bootstrap.rs

```diff
@@ -567,13 +518,111 @@ fn node_to_config(node: &Node) -> NodeConfiguration {
+// we need to configure and push a SetupOS image for each node.
+pub fn setup_nested_vms(
+    nodes: &[NestedNode],
+    env: &TestEnv,
+    farm: &Farm,
+    group_name: &str,
+    nns_url: &Url,
+    nns_public_key: &str,
+    to: &NestedVersionTarget,
+) -> anyhow::Result<()> {
+    let mut join_handles: Vec<JoinHandle<anyhow::Result<()>>> = vec![];
+    for node in nodes {
+        let t_farm = farm.to_owned();
+        let t_env = env.to_owned();
+        let t_group_name = group_name.to_owned();
+        let t_vm_name = node.name.to_owned();
+        let t_nns_url = nns_url.to_owned();
+        let t_nns_public_key = nns_public_key.to_owned();
+        let t_to = to.to_owned();
+        join_handles.push(thread::spawn(move || {
+            let configured_image =
+                configure_setupos_image(&t_env, &t_vm_name, &t_nns_url, &t_nns_public_key, &t_to)?;
+
+            let configured_image_spec = AttachImageSpec::new(t_farm.upload_file(
+                &t_group_name,
+                configured_image,
+                NESTED_CONFIGURED_IMAGE_PATH,
+            )?);
+            t_farm.attach_disk_images(
+                &t_group_name,
+                &t_vm_name,
+                "usb-storage",
+                vec![configured_image_spec],
+            )?;
+
+            Ok(())
+        }));
+    }
+
+    let mut result = Ok(());
+    // Wait for all threads to finish and return an error if any of them fails.
+    for jh in join_handles {
+        if let Err(e) = jh.join().expect("waiting for a thread failed") {
+            warn!(farm.logger, "setting up VM failed with: {:?}", e);
+            result = Err(anyhow::anyhow!("failed to set up a VM pool"));
+        }
+    }
+
+    result
+}
+
+pub fn start_nested_vms(env: &TestEnv, farm: &Farm, group_name: &str) -> anyhow::Result<()> {
+    for node in env.get_all_nested_vms()? {
+        let t_farm = farm.to_owned();
+        let t_vm_name = node.vm_name().to_owned();
+        let t_group_name = group_name.to_owned();
+
+        t_farm.start_vm(&t_group_name, &t_vm_name)?;
+    }
+
+    Ok(())
+}
+
+pub fn configure_setupos_image(
     env: &TestEnv,
     name: &str,
     nns_url: &Url,
     nns_public_key: &str,
+    to: &NestedVersionTarget,
 ) -> anyhow::Result<PathBuf> {
-    let setupos_image = env.get_dependency_path("ic-os/setupos/envs/dev/disk-img.tar.zst");
+    let tmp_dir = env.get_path("setupos");
+    fs::create_dir_all(&tmp_dir)?;
+
+    let setupos_image = match to {
+        NestedVersionTarget::Branch(false) => {
+            env.get_dependency_path("ic-os/setupos/envs/dev/disk-img.tar.zst")
+        }
+        NestedVersionTarget::Branch(true) => {
+            bail!("Starting nested VMs from `-test` versions is unsupported.");
+        }
+        NestedVersionTarget::Mainnet => {
+            let url = env.get_mainnet_setupos_img_url()?;
+
+            download_setupos_image(&tmp_dir, &url, None)?
```

Comment:
Why not download this via Bazel such that it will be cached like we do for [the mainnet NNS and SNS canisters](https://sourcegraph.com/github.com/dfinity/ic/-/blob/mainnet-canisters.bzl)?